clippy_lints/src/use_self.rs

```diff
@@ -81,192 +70,220 @@ fn span_use_self_lint(cx: &LateContext<'_, '_>, path: &Path<'_>, last_segment: O
     );
 }
 
-// FIXME: always use this (more correct) visitor, not just in method signatures.
-struct SemanticUseSelfVisitor<'a, 'tcx> {
+fn span_lint_ignore_last_segment<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, path: &'tcx Path<'tcx>) {
+    if path.segments.len() > 1 {
+        span_lint(cx, truncate_last_segment(cx, path.span))
+    }
+}
+
+fn span_lint_on_qpath_resolved<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, qpath: &'tcx QPath<'tcx>, enum_variant: bool) {
+    match qpath {
+        QPath::Resolved(_, path) => {
+            if enum_variant {
+                span_lint_ignore_last_segment(cx, path);
+            } else {
+                span_lint(cx, path.span);
+            }
+        },
+        _ => (),
+    };
+}
+
+fn truncate_last_segment<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, sp: Span) -> Span {
+    match snippet_opt(cx, sp) {
+        Some(snippet) => match snippet.rfind("::") {
+            Some(bidx) => sp.with_hi(sp.lo() + BytePos(bidx as u32)),
+            None => sp,
+        },
+        None => sp,
+    }
+}
```

Comment:
This seems unwise when paths can have generic args anywhere. You probably want to instead rely on the fact that segments have their own spans (AFAIK) so you can check that the path's overall span and the span of the last segment have the same `hi` and then use that to "carve out" the last segment.

Reply:
I've tried to depend more on paths subelements own span. I had issues with `QPath::TypeRelative`. @flip1995 spotted issues in the ast lowering. See the associated comment in the patch line 172 and the discussion in discord.

But agreed that this will fail upon generics nested paths. I'll add a test case and look for options to deal with that.

Reply:
Span handling improved. See `span_lint_until_last_segment`. I added a nested path test case. It fails to compile due to the reason @flip1995 analyzed above.